### ISSUE TITLE
CodeGenerators? Code Fix for RS1024 is horribly wrong when multiline strings are present

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#56165]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/CodeGenerators-Code-Fix-for-RS1024-is-h/1517562)._

---
In my code generator, I appear to automatically receive references to these analyzers:
```
Microsoft.CodeAnalysis.Analyzers
Microsoft.CodeAnalysis.CSharp.Analyzers
```

In this project, I get warning `RS1024` for the snippet `Static.GetHashCode("")` in the following code:
```
public class Static
{
	public static int GetHashCode(string s) => 0xabcdef8;
}
public class RS1024CodeFixIssue
{
	public static string GetMultiLine() => $@"
HashCode 0x{Static.GetHashCode(""):x8}
";
}
```

After attempting to fix this with the light bulb, the code looks as follows:
```
	public static string GetMultiLine() => $@"
#pragma warning disable RS1024 // Compare symbols correctly

HashCode 0x{Static.GetHashCode(""):x8}

#pragma warning restore RS1024 // Compare symbols correctly
";
```

Obviously, RS1024 is still there.

---
#### Feedback Bot on 8/31/2021, 01:28 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by Youssef1313** [comment]
The issue isn't specific to RS1024, and this is not even the codefix of RS1024.

I think the fix will be to override the virtual [`GetAdjustedTokenForPragmaXXX`](https://github.com/dotnet/roslyn/blob/a657a11b4bcf7141938260688fccae223bf7a206/src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.cs#L90-L94) for C# to get out of an interpolated string, and tweak the [VB overrides](https://github.com/dotnet/roslyn/blob/a657a11b4bcf7141938260688fccae223bf7a206/src/Features/VisualBasic/Portable/CodeFixes/Suppression/VisualBasicSuppressionCodeFixProvider.vb#L73-L103).

**Comment by ericwj** [comment]
> The issue isn't specific to RS1024

Sure I suspected that. It just happens to be the first time that I ran into it.

> and this is not even the codefix of RS1024

How do you mean, seeing the code that is inserted mentions the exact error code?

I could also argue the warning is wrong, because all I did was declare something called `GetHashCode` and trying to dump the resulting value into a string. I shouldn't get any warning for that.

**Comment by Youssef1313** [comment]
> How do you mean, seeing the code that is inserted mentions the exact error code?

This is a generic codefix that suppresses almost any configurable diagnosticl The codefix for RS1024 is the one that corrects the code to use SymbolEqualityComparer.

---

> I could also argue the warning is wrong, because all I did was declare something called GetHashCode and trying to dump the resulting value into a string. I shouldn't get any warning for that.

For this, it would be an issue for dotnet/roslyn-analyzers repository. I recall that @ryzngard has fixed issues for RS1024, but I don't know if this is one of them. @ryzngard @mavasani @jmarolf A new version of Microsoft.CodeAnalysis.Analyzers would be nice. The latest version is 9 months old.

**Comment by ryzngard** [comment]
I put in this repo because the issue is with the "suppress" not working correctly in an interpolated string. That logic, which adds the `pragma warning disable` is in this repo iirc. Agree that this is a generic issue and not specific to RS1024. However, the RS1024 issue is fixed with https://github.com/dotnet/roslyn-analyzers/pull/5057

**Comment by Youssef1313** [comment]
> I put in this repo because the issue is with the "suppress" not working correctly in an interpolated string

Yes that's correct. That's what I was trying to explain in https://github.com/dotnet/roslyn/issues/56165#issuecomment-912410509.
